range_queries/heavy_light_decomposition.cpp

```diff
@@ -0,0 +1,598 @@
+   * @param node the node where the update is done
+   * @param val the value that is being updated
+   * @returns void
+   */
+  void update(int node, X val) {
+    X diff = val - Tree<X>::t_val[node];
+    SG<X>::update(h_label[node], diff);
+    Tree<X>::t_val[node] = val;
+  }
+
+  /**
+   * @brief This function returns the sum of node values in the simple path from
+   * from node_1 to node_2
+   * @param a the node where the simple path starts
+   * @param b the node where the simple path ends
+   * (parameters are interchangeable, i.e., the function is commutative)
+   * @returns the sum of node values in the simple path from a to b
+   */
+  X query(int a, int b) {
+    int lc = Tree<X>::lca(a, b);
+    X ret = SG<X>::sret_init;
+    assert(lc != -1);
+    ret += chain_query(a, lc);
+    ret += chain_query(b, lc);
+    return ret - Tree<X>::t_val[lc];
+  }
+};
+} // namespace heavy_light_decomposition
+} // namespace range_queries
+
+/**
+ * Test implementations
+ * @returns none
+ */
+static void test_1() {
+  std::cout << "Test 1:\n";
+
+  // Test details
+  int n = 5;
+  std::vector<int64_t> node_values = {4, 2, 5, 2, 1};
+  std::vector<std::vector<int>> edges = {{1, 2}, {1, 3}, {3, 4}, {3, 5}};
+  std::vector<std::vector<int>> queries = {
+      {2, 4},
+      {1, 3, 2},
+      {2, 4},
+  };
+
+  range_queries::heavy_light_decomposition::HLD<int64_t> hld(n);
+  hld.set_node_val(node_values);
+  for (int i = 0; i < n - 1; i++) {
+    int u = edges[i][0], v = edges[i][1];
+    hld.add_edge(u - 1, v - 1);
+  }
+  hld.init();
+  for (const auto &q : queries) {
+    int type = q[0];
+    if (type == 1) {
+      int p = q[1], x = q[2];
+      hld.update(p - 1, x);
+    } else if (type == 2) {
+      int p = q[1];
+      std::cout << hld.query(0, p - 1) << std::endl;
+    } else {
+      continue;
+    }
+  }
+  std::cout << "\nTest 1 passed!\n";
+}
+
+/**
+ * Second test implementations
+ * @returns void
+ */
+static void test_2() {
+  std::cout << "Test 2:\n";
+
+  // Test details (Bamboo)
+  int n = 10;
+  std::vector<int64_t> node_values = {1, 8, 6, 8, 6, 2, 9, 2, 3, 2};
+  std::vector<std::vector<int>> edges = {
+      {10, 5}, {6, 2}, {10, 7}, {5, 2}, {3, 9}, {8, 3}, {1, 4}, {6, 4}, {8, 7}};
+  std::vector<std::vector<int>> queries = {
+      {2, 10},   {2, 6}, {1, 3, 4}, {2, 9},    {1, 5, 3},
+      {1, 7, 8}, {2, 4}, {2, 8},    {1, 1, 4}, {1, 2, 7}};
+
+  range_queries::heavy_light_decomposition::HLD<int64_t> hld(n);
+  hld.set_node_val(node_values);
+  for (int i = 0; i < n - 1; i++) {
+    int u = edges[i][0], v = edges[i][1];
+    hld.add_edge(u - 1, v - 1);
+  }
+  hld.init();
+  for (const auto &q : queries) {
+    int type = q[0];
+    if (type == 1) {
+      int p = q[1], x = q[2];
+      hld.update(p - 1, x);
+    } else if (type == 2) {
+      int p = q[1];
+      std::cout << hld.query(0, p - 1) << std::endl;
+    } else {
+      continue;
+    }
+  }
+  std::cout << "\nTest2 passed!\n";
```

Comment:
In order to say that the test passed, shouldn't there be some sort of a check? In the repo, we use `assert()` to ensure that the values are as expected. Refer to the Fenwick Tree code, for example.